File: src/main/java/org/radarbase/appserver/service/questionnaire/schedule/ProtocolHandlerRunner.java
package org.radarbase.appserver.service.questionnaire.schedule;

import org.radarbase.appserver.dto.protocol.Assessment;
import org.radarbase.appserver.dto.questionnaire.AssessmentSchedule;
import org.radarbase.appserver.entity.User;
import org.radarbase.appserver.service.questionnaire.protocol.ProtocolHandler;

import java.util.ArrayList;
import java.util.List;

@SuppressWarnings("PMD.DataflowAnomalyAnalysis")
public class ProtocolHandlerRunner {
    private transient List<ProtocolHandler> protocolHandlers = new ArrayList<>();

    public AssessmentSchedule runProtocolHandlers(Assessment assessment, User user) {
        AssessmentSchedule assessmentSchedule = new AssessmentSchedule();
        for (ProtocolHandler leaf : this.protocolHandlers) {
            assessmentSchedule = leaf.handle(assessmentSchedule, assessment, user);
        }
        return assessmentSchedule;
    }

    public void addProtocolHandler(ProtocolHandler protocolHandler) {
        if (protocolHandler != null) this.protocolHandlers.add(protocolHandler);
    }
}

File: src/main/java/org/radarbase/appserver/service/questionnaire/schedule/ProtocolHandlerRunner.kt
package org.radarbase.appserver.service.questionnaire.schedule

import org.radarbase.appserver.dto.protocol.Assessment
import org.radarbase.appserver.dto.questionnaire.AssessmentSchedule
import org.radarbase.appserver.entity.User
import org.radarbase.appserver.service.questionnaire.protocol.ProtocolHandler

class ProtocolHandlerRunner {
    @Transient
    private val protocolHandlers: MutableList<ProtocolHandler> = ArrayList<ProtocolHandler>()

    fun runProtocolHandlers(assessment: Assessment, user: User): AssessmentSchedule {
        var assessmentSchedule = AssessmentSchedule()
        for (leaf in this.protocolHandlers) {
            assessmentSchedule = leaf.handle(assessmentSchedule, assessment, user)
        }
        return assessmentSchedule
    }

    fun addProtocolHandler(protocolHandler: ProtocolHandler?) {
        if (protocolHandler != null) this.protocolHandlers.add(protocolHandler)
    }
}
